### ISSUE TITLE
/chain/block/:blockid API fails on the block that has Deploy transaction.

### ISSUE THREAD

**Issue by mrshah-at-work** [hyperledger-archives/fabric#2123]
## Description

`/chain/block` API fails on the block that has Deploy transaction.
It returns the following error:
`{"Error": "proto: protos.ChaincodeDeploymentSpec: wiretype end group for non-group"}`
## Describe How to Reproduce

1. Create a network with 4 Peers & Membership services (I am running with TLS turned on, but this is optional).
2. Deploy the example02 chaincode on peer 1.
3. Wait for 90 seconds for the chaincode container to come up.
   NOTE: The block height is increased by 1.
4. Execute a few transactions (Invokes & Queries).
   NOTE: Invokes will increase the block height.
5. Call the chain/block API:
   `/chain/blocks/0` - Success
   `/chain/blocks/1` - Failure (Block with Deploy transaction)
   `/chain/blocks/2` - Success
   `/chain/blocks/3` - Success
6. Continue with another deploy transaction (Marbles).
7. Call the chain/block API:
   `/chain/blocks/0` - Success
   `/chain/blocks/1` - Failure (Block with Deploy transaction)
   `/chain/blocks/2` - Success
   `/chain/blocks/3` - Success
   `/chain/blocks/4` - Failure (Block with Deploy transaction)

**Comment by srderson** [comment]
is this on the master branch or 0.5 branch?

**Comment by mrshah-at-work** [comment]
I am sorry I missed it. 
I should have mentioned it here - this is on `v0.5-developer-preview` branch

**Comment by srderson** [comment]
Does the gRPC API work in your environment for deploying transactions? I'm wondering if this is specific to the REST API or a larger issue?

**Comment by mrshah-at-work** [comment]
The deploy using the REST API is successful and all the other requests work fine. I am able to query and invoke chaincode successfully.

**Comment by muralisrini** [comment]
@mrshah-at-ibm I couldn't recreate on latest (see chain/blocks/1 below) using the following
- security with privacy
- noops
- latest code
- vagrant env

```
Deploy and an Invoke
---------------------
CORE_SECURITY_ENABLED=true CORE_SECURITY_PRIVACY=true ./peer chaincode deploy -p github.com/hyperledger/fabric/examples/chaincode/go/chaincode_example02 -c '{"Function":"init","Args":["a","100","b","200"]}' -u lukas
CORE_SECURITY_ENABLED=true CORE_SECURITY_PRIVACY=true ./peer chaincode invoke -n a5389f7dfb9efae379900a41db1503fea2199fe400272b61ac5fe7bd0c6b97cf10ce3aa8dd00cd7626ce02f18accc7e5f2059dae6eb0786838042958352b89fb -c '{"Function":"invoke","Args":["a","10"]}' -u lukas

curl for blocks
---------------
curl http://localhost:5000/chain/blocks/0
{"nonHashData":{"localLedgerCommitTimestamp":{"seconds":1467919685,"nanos":71643416}}}

curl http://localhost:5000/chain/blocks/1
{"transactions":[{"type":1,"chaincodeID":"CkdnaXRodWIuY29tL2h5cGVybGVkZ2VyL2ZhYnJpYy9leGFtcGxlcy9jaGFpbmNvZGUvZ28vY2hhaW5jb2RlX2V4YW1wbGUwMhKAAWE1Mzg5ZjdkZmI5ZWZhZTM3OTkwMGE0MWRiMTUwM2ZlYTIxOTlmZTQwMDI3MmI2MWFjNWZlN2JkMGM2Yjk3Y2YxMGNlM2FhOGRkMDBjZDc2MjZjZTAyZjE4YWNjYzdlNWYyMDU5ZGFlNmViMDc4NjgzODA0Mjk1ODM1MmI4OWZi","payload":"CvIBCAESzAEKR2dpdGh1Yi5jb20vaHlwZXJsZWRnZXIvZmFicmljL2V4YW1wbGVzL2NoYWluY29kZS9nby9jaGFpbmNvZGVfZXhhbXBsZTAyEoABYTUzODlmN2RmYjllZmFlMzc5OTAwYTQxZGIxNTAzZmVhMjE5OWZlNDAwMjcyYjYxYWM1ZmU3YmQwYzZiOTdjZjEwY2UzYWE4ZGQwMGNkNzYyNmNlMDJmMThhY2NjN2U1ZjIwNTlkYWU2ZWIwNzg2ODM4MDQyOTU4MzUyYjg5ZmIaFgoEaW5pdBIBYRIDMTAwEgFiEgMyMDAqBWx1a2FzMAE=","uuid":"a5389f7dfb9efae379900a41db1503fea2199fe400272b61ac5fe7bd0c6b97cf10ce3aa8dd00cd7626ce02f18accc7e5f2059dae6eb0786838042958352b89fb","timestamp":{"seconds":1467922219,"nanos":361601500}}],"stateHash":"E8yUgvFk6FsZkQY2g9OhBemcqsxqK9PpHRe71Hd3oaciujcRxJF3Cnm0OkXnc/DHSqc/2/lkLDhkCGNpurHNEg==","previousBlockHash":"RrndKwuojRMjOz/rdD7rJD/NUupiuBuCtQwnZG7Vdi/XXcTd2MDyAMsFAZ1ntZL2/IIcSUeatIZAKS6ss7fEvg==","nonHashData":{"localLedgerCommitTimestamp":{"seconds":1467922232,"nanos":659644181}}}

curl http://localhost:5000/chain/blocks/2
{"stateHash":"E8yUgvFk6FsZkQY2g9OhBemcqsxqK9PpHRe71Hd3oaciujcRxJF3Cnm0OkXnc/DHSqc/2/lkLDhkCGNpurHNEg==","previousBlockHash":"FNJWlrXVRbkYDY1PhZf5khYa73cECT4BPFMC2QlR0JyW4ijoHhmFSaFcvSG+Lp/Tv4dHlIPwkNDB+TNzssC8mQ==","nonHashData":{"localLedgerCommitTimestamp":{"seconds":1467922389,"nanos":529854784}}}

```

**Comment by mrshah-at-work** [comment]
@muralisrini are you using the `0.5-developer-preview` branch?

**Comment by muralisrini** [comment]
@mrshah-at-ibm No, I'll try that branch

**Comment by muralisrini** [comment]
@mrshah-at-ibm I can recreate the issue in `0.5-developer-preview` branch

**Comment by mrshah-at-work** [comment]
We should also bring it to @angrbrd's attention - tagging her here.

**Comment by muralisrini** [comment]
actually can recreate in both branches... I must not have tested correctly in master ... or the failure does not happen always. This is not a regression (shoud have been there forever).

Issue around this area - https://github.com/hyperledger/fabric/blob/master/core/rest/api.go#L107

The issue is same as what we ran into when sending block events when privacy is enabled - viz, attempt to remove the large payload from deploy transactions can fail if the unmarshal fails. In the case of secure transactions it'll likely fail (use the word "likely" because I wonder if its possible for the unmarshall to wrongly succeed sometimes).

A reasonable course of action - given we cannot decrypt the transaction -  is to conditionally ignore the failure and set the deploy payload to nil as opposed to returning the error.

One could argue that given this is a benign error, user just ignores it. However two problems with that argument
- we can't distinguish between a real error and a benign error
- a single failing deploy transaction in a block will prevent information on all transactions in the block (due to the error bailing out)

@adecaro what would it take to fix this right ? Actually decrypt the transaction so we can set payload to nil. This is probably something we'll have to deal with soon anyway even if we just go with the workaround for this bug.

**Comment by mastersingh24** [comment]
@muralisrini - are you just using privacy as an example here?  I am pretty sure that in @mrshah-at-ibm 's case we don't use privacy

**Comment by muralisrini** [comment]
@mastersingh24 : I used privacy as that's the only way I could recreate the issue.

Also, I don't need the peer in privacy mode. Just issuing the deploy with privacy would do the trick.

So for example following works

```
CORE_SECURITY_ENABLED=true CORE_LOGGING_LEVEL=debug ./peer node start 

./peer network login lukas

CORE_SECURITY_ENABLED=true ./peer chaincode deploy -p github.com/hyperledger/fabric/examples/chaincode/go/chaincode_example02 -c '{"Function":"init","Args":["a","100","b","200"]}' -u lukas

curl http://localhost:5000/chain/blocks/1
{"transactions":[{"type":1,"chaincodeID":"Ckd....


```

We get the  identical error if we add CORE_SECURITY_PRIVACY=true to the `deploy` (keeping it false on the peer). So is it possible the deploy is issued with security.privacy=true ?

```

CORE_SECURITY_ENABLED=true CORE_SECURITY_PRIVACY=true ./peer chaincode deploy -p github.com/hyperledger/fabric/examples/chaincode/go/chaincode_example02 -c '{"Function":"init","Args":["a","100","b","200"]}' -u lukas

curl http://localhost:5000/chain/blocks/2
{"Error":"proto: protos.ChaincodeDeploymentSpec: wiretype end group for non-group"}

```

@mrshah-at-ibm ?

**Comment by mrshah-at-work** [comment]
@muralisrini & @mastersingh24 
yep, PRIVACY is set to `true` in our situation.

**Comment by muralisrini** [comment]
@mrshah-at-ibm  PRIVACY set in peer startup _and_ for deploy/invoke ? If you don't have it peer startup, do you need it for deploy/invoke ?

**Comment by mrshah-at-work** [comment]
We have it in peer startup - `CORE_SECURITY_PRIVACY=true`

**Comment by adecaro** [comment]
@muralisrini, sorry for the late reply. 

Given the nature of the query, the validator should not decrypt the payload of the confidential deploy transactions because the validator does not have a way to understand if the invoker has the right to access that data. Indeed, if the invoker has this right, meaning he has a secret key, he will just decrypt the payload by himself.

Then, setting the payload field to nil is fine if we want to save bandwidth. There should a way then to retrieve the payload of specific transaction then, no?

**Comment by muralisrini** [comment]
@adecaro thanks for the feedback. I'll put out a PR to just set the payload to nil for the deploy transaction. This is better than suppressing information on all transactions in the block due to failure to decrypt one deploy transaction.

will look to @angrbrd to review please.

**Comment by mrshah-at-work** [comment]
@tamasblummer @muralisrini  can we also add this fix to v0.5-developer-preview branch?

**Comment by muralisrini** [comment]
@mrshah-at-ibm @tamasblummer  also submitted #2173 for the dev preview branch... but I see travis failed (it shouldn't have...).

**Comment by srderson** [comment]
This has been resolved. The fix was merged into both the master and 0.5 branch and master branch.
